pytest_django/plugin.py

```diff
@@ -514,6 +514,20 @@ def _template_string_if_invalid_marker(request):
             else:
                 dj_settings.TEMPLATE_STRING_IF_INVALID.fail = False
 
+
+@pytest.fixture(autouse=True, scope='function')
+def django_clear_site_cache():
+    """Clears ``django.contrib.sites.models.SITE_CACHE`` to avoid
+    unexpected behavior when cached site objects.
+    """
+
+    if django_settings_is_configured():
+        from django.conf import settings as dj_settings
+
+        if 'django.contrib.sites' in dj_settings.INSTALLED_APPS:
+            from django.contrib.sites import models
+            models.SITE_CACHE.clear()
```

Comment:
I think the benefit of this autouse fixture overcomes the overhead of a list lookup, in comparison with all setup/teardown that is done with databases etc. 

We could cache the list lookup, but is that really worth it? Can we guarantee that INSTALLED_APPS are not modified by tests?

The core reason for doing this, is that the SITE_CACHE is evil :), and people that write tests for their projects where sites are used, they will tear their hair off when site objects suddenly starts to mismatch with referenced objects. The SITE_CACHE is nice for production, but not for testing...

I was against having the mailbox cleaned with an autouse fixture, as we talked about in #410. But this is a different thing, this is a cache that sits there in its evil darkness and just makes people that loves testing wanna kill them self when debugging strange errors that the SITE_CACHE causes...